h2o-core/src/main/java/water/Lockable.java

```diff
@@ -113,7 +113,7 @@ public final Futures delete(Key<Job> job_key, Futures fs, boolean cascade) {
     @Override public Lockable atomic(Lockable old) {
       _old = old;
       if( old != null ) {       // Prior Lockable exists?
-        assert !old.is_wlocked(_job_key) : "Key "+_key+" already locked (or deleted); lks="+Arrays.toString(old._lockers); // No double locking by same job
+        if(old.is_wlocked(_job_key)) return old;
```

Comment:
@michalkurka can we not fail if Lockable is already locked with the same key? This way we can probably include `write_lock` inside `update`. I don't know if there are situations when we want to update without prior locking. Maybe local updates do not require distributed locks( then it is better not to include write_lock inside update to give flexibility) .
Of course without failing developers are free to abuse this ability and use write_lock unnecessary number of times. But still it is easier than trying to understand why some execution paths have different lock story. I run into such situation : for GBM model parameters process of building a model does not release lock but for TargetEncoder model parameters it does. But it is hard to understand who and where releases the lock in TE path.

Reply:
this is outdated. But still interesting to know.